### DISCUSSION TITLE
Use motion capture with crazyflie-lib-python

### DISCUSSION THREAD

**Discussion by cc299792458** [bitcraze/discussions#240] [Q&A] [answered]
Hi, Bitcraze!

I'm sorry to bother you but I have some problems when I use motion caputure(VICON) with crazyflie-lib-python. I'm also confused if this is available.
I think the main step is sending the data gets from mocap to crazyflie. So I've just tried a sample demo but it doesn't work.
The demo is:
(1)I use crazyflie.extpose.send_pose() to send the data(which I set x=0, y=0, z=0, qx=0, qy=0, qz=0, qw=1, just used to test) to my crazyflie.
(2)I use Logger(implied by LogConfig) to check the position and rotation of it.
I think if position and rotation are: x=0, y=0, z=0, roll=0, pitch=0, yaw=0, that means I successed.
But I can't see the change after I do so. Data print in step(2) has no difference between when I check it without step (1).

Meanwhile, I also encounter a strange thing that when I do step(2) both with and without step(1), the x value and y value equals to 0 all the time, even if I change the position of crazyflie in the real world manually.

So my question is:
1. If I can use motion capture with crazyflie-lib-python in this way?
2. Is there any mistake in my demo?
3. I wonder what's the reason of the "strange thing", is that normal or not?

Thanks in advance!

**Comment by krichardsson** [comment, marked as answer]
It sounds like you are on the right track

I don't think `crazyflie.extpose.send_pose()` exists but you probably mean `cf.extpos.send_extpose()` :-)

crazyflie.extpose.send_extpose() pushes your data to the kalman estimator, pretty much like any other sensor data and if you do it repeatedly it should work as you expect.

> If I can use motion capture with crazyflie-lib-python in this way?

Yes. We don't have a Vicon example, but we have one for [Qualisys](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/qualisys/qualisys_hl_commander.py).  [Lines 228-245](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/qualisys/qualisys_hl_commander.py#L228-L245) is where the pose data is sent to the Crazyflie. Also note that you can use `cf.extpos.send_extpos()` if you only want to send position data and not full pose. 

> Is there any mistake in my demo?

The two things I can think about is that you perhaps did not [activate the kalman estimator](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/qualisys/qualisys_hl_commander.py#L257-L262), or that you do not call `send_extpose()` repeatedly?

> I wonder what's the reason of the "strange thing", is that normal or not?

If you only log (not pushing anything to `send_extpose()`) I would expect x, y and z to change. You can use the client to set up logging of the variables to see what happens when you move around, just to rule out programming errors in your code.

**Reply by cc299792458** [reply]
Thank you very much!
As you mentioned, I think probably I forgot to activate the kalman estimator and used the default one(complementary filters).
I will try it later and also use the client to check variables, Thank you for your suggestions!
By the way, you also suggest that I can use cf.extpos.send_extpos() if I only want to send position data. I remember someone said they got into trouble(crazyflie became crazy, etc) when they used cf.extpos.extpose() (but maybe happened a long time ago). May I ask, if I use cf.extpos.send_extpose() correctly, I can send the full pose to crazyflie, right?

**Reply by krichardsson** [reply]
`send_extpose()` should work fine, but if I remember correctly it is a bit sensitive to noise (mainly for orientation) when the yaw is close to +-90 degrees. When the data is too noisy it has a tendency of "going bananas" (diverging).

**Reply by cc299792458** [reply]
Get, thanks!

**Comment by krichardsson** [comment]
I just wanted to add that I recently found a bug in the [qualisys_hl_commander.py](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/qualisys/qualisys_hl_commander.py) that has been fixed now. I also added functionality to tune down the sensitivity of the orientation information that is pushed to the estimator.

**Comment by mbaytas** [comment]
We recently published the open source Qualisys Drone SDK for controlling the Crazyflie (and soon, hopefully, the Robomaster TT) with Qualisys motion capture systems: https://github.com/qualisys/qualisys_drone_sdk

The SDK is not designed to be universal, it is specific to Qualisys systems. However, you might find inspiration in our code, and adapt it to your own application @cc299792458.

**Reply by cc299792458** [reply]
How nice of you to remind! I will check it, thank you!
